### ISSUE TITLE
arc-runner-set not scaling down intermittently | gha-runner-scale-set:0.6.1

### ISSUE THREAD

**Issue by ChaitanyaAtchuta5** [actions/actions-runner-controller#2941]
### Checks

- [X] I've already read https://github.com/actions/actions-runner-controller/blob/master/TROUBLESHOOTING.md and I'm sure my issue is not covered in the troubleshooting guide.
- [X] I'm not using a custom entrypoint in my runner image

### Controller Version

0.6.1

### Helm Chart Version

0.6.1

### Deployment Method

Helm

### cert-manager installation

NA

### Checks

- [X] This isn't a question or user support case (For Q&A and community support, go to [Discussions](https://github.com/actions/actions-runner-controller/discussions). It might also be a good idea to contract with any of contributors and maintainers if your business is so critical and therefore you need priority support
- [X] I've read [releasenotes](https://github.com/actions/actions-runner-controller/tree/master/docs/releasenotes) before submitting this issue and I'm sure it's not due to any recently-introduced backward-incompatible changes
- [X] My actions-runner-controller version (v0.x.y) does support the feature
- [X] I've already upgraded ARC (including the CRDs, see charts/actions-runner-controller/docs/UPGRADING.md for details) to the latest and it didn't fix the issue
- [X] I've migrated to the workflow job webhook event (if you using webhook driven scaling)

### Resource Definitions

```yaml
template:  
  spec:
    containers:
      - name: runner
        image: ghcr.io/actions/actions-runner:latest
        #intentionally kept 10s sleep to make sure istio-proxy container is up and running
        command: [ "/bin/bash","-c","sleep 10 && /home/runner/run.sh" ]
```

### To Reproduce

```markdown
I am trying out github ARC
1. I have created a sample workflow which just has 1 job and performs sleep 30s and prints Hello World.
2. I have managed to set-up the arc-gha-rs-controller and arc-runner-set properly and they are up and running.
3. Job is running as expected but some times after the job completion the runner-set is not scaling down to 0. (Intermittent)
```

### Describe the bug

When the job finishes runner-scale-set is not scaling down to 0. This is a intermittent behavior. So I have this stale runner staying there in the cluster(I have not set any min or max runners). Then if I trigger a new job this stale runner is not picking up the job and job will be in queue state for ever.
<img width="1086" alt="image" src="https://github.com/actions/actions-runner-controller/assets/68017180/7e2f15d6-62b3-4dda-9d30-bc6972e9869c">

### Describe the expected behavior

As soon as the job completes the runner should terminate.

### Whole Controller Logs

```shell
https://gist.github.com/ChaitanyaAtchuta5/7692ccd1e35e4b6706f8a0f20a570aaf
```

### Whole Runner Pod Logs

```shell
Logs when not working as expected (Runner pod not getting terminated when job completes)
https://gist.github.com/ChaitanyaAtchuta5/881715bfec200f42de97adebec44e926
Logs when working as expected (Runner pod gets terminated when job completes)
https://gist.github.com/ChaitanyaAtchuta5/c847653cb266dcc1b010a427767a2d51
```

### Additional Context

<img width="1660" alt="image" src="https://github.com/actions/actions-runner-controller/assets/68017180/5ee68900-0372-40b6-9c2a-ebd86647795f">

**Comment by Link-** [comment]
From the controller logs. It looks like you uninstalled the resources so the k8s secret was deleted before the controller got the chance to delete the runners from the Actions backend service.

```
2023-09-27T03:37:05Z	INFO	EphemeralRunnerSet	Waiting for resources to be deleted	{"ephemeralrunnerset": "github-runners/arc-runner-set-b2xp4"}
2023-09-27T03:37:05Z	INFO	Starting workers	{"controller": "ephemeral-runner-controller", "controllerGroup": "actions.github.com", "controllerKind": "EphemeralRunner", "worker count": 1}
2023-09-27T03:37:05Z	INFO	Starting workers	{"controller": "autoscalingrunnerset", "controllerGroup": "actions.github.com", "controllerKind": "AutoscalingRunnerSet", "worker count": 1}
2023-09-27T03:37:05Z	INFO	Starting workers	{"controller": "autoscalinglistener", "controllerGroup": "actions.github.com", "controllerKind": "AutoscalingListener", "worker count": 1}
2023-09-27T03:37:05Z	INFO	AutoscalingRunnerSet	Deleting resources	{"autoscalingrunnerset": "github-runners/arc-runner-set"}
2023-09-27T03:37:05Z	INFO	AutoscalingRunnerSet	Cleaning up the listener	{"autoscalingrunnerset": "github-runners/arc-runner-set"}
2023-09-27T03:37:05Z	INFO	AutoscalingRunnerSet	Listener is deleted	{"autoscalingrunnerset": "github-runners/arc-runner-set"}
2023-09-27T03:37:05Z	INFO	AutoscalingRunnerSet	Cleaning up ephemeral runner sets	{"autoscalingrunnerset": "github-runners/arc-runner-set"}
2023-09-27T03:37:05Z	INFO	AutoscalingRunnerSet	Deleting all ephemeral runner sets	{"autoscalingrunnerset": "github-runners/arc-runner-set", "count": 1}
2023-09-27T03:37:05Z	INFO	AutoscalingRunnerSet	Skip ephemeral runner set since it is already marked for deletion	{"autoscalingrunnerset": "github-runners/arc-runner-set", "name": "arc-runner-set-b2xp4"}
2023-09-27T03:37:05Z	INFO	AutoscalingRunnerSet	Waiting for ephemeral runner sets to be deleted	{"autoscalingrunnerset": "github-runners/arc-runner-set"}
2023-09-27T03:37:05Z	INFO	AutoscalingRunnerSet	Deleting resources	{"autoscalingrunnerset": "github-runners/arc-runner-set-sample"}
2023-09-27T03:37:05Z	INFO	AutoscalingRunnerSet	Cleaning up the listener	{"autoscalingrunnerset": "github-runners/arc-runner-set-sample"}
2023-09-27T03:37:05Z	INFO	AutoscalingRunnerSet	Listener is deleted	{"autoscalingrunnerset": "github-runners/arc-runner-set-sample"}
2023-09-27T03:37:05Z	INFO	AutoscalingRunnerSet	Cleaning up ephemeral runner sets	{"autoscalingrunnerset": "github-runners/arc-runner-set-sample"}
2023-09-27T03:37:05Z	INFO	AutoscalingRunnerSet	All ephemeral runner sets are deleted	{"autoscalingrunnerset": "github-runners/arc-runner-set-sample"}
2023-09-27T03:37:05Z	INFO	AutoscalingRunnerSet	Skipping cleaning up kubernetes mode service account	{"autoscalingrunnerset": "github-runners/arc-runner-set-sample", "reason": "annotation key \"actions.github.com/cleanup-kubernetes-mode-role-binding-name\" not present"}
2023-09-27T03:37:05Z	INFO	AutoscalingRunnerSet	Skipping cleaning up kubernetes mode role	{"autoscalingrunnerset": "github-runners/arc-runner-set-sample", "reason": "annotation key \"actions.github.com/cleanup-kubernetes-mode-role-name\" not present"}
2023-09-27T03:37:05Z	INFO	AutoscalingRunnerSet	Skipping cleaning up kubernetes mode role binding	{"autoscalingrunnerset": "github-runners/arc-runner-set-sample", "reason": "annotation key \"actions.github.com/cleanup-kubernetes-mode-service-account-name\" not present"}
2023-09-27T03:37:05Z	INFO	AutoscalingRunnerSet	Removing finalizer from no permission service account	{"autoscalingrunnerset": "github-runners/arc-runner-set-sample", "name": "arc-runner-set-sample-gha-rs-no-permission"}
2023-09-27T03:37:05Z	INFO	AutoscalingRunnerSet	No permission service account has already been deleted	{"autoscalingrunnerset": "github-runners/arc-runner-set-sample", "name": "arc-runner-set-sample-gha-rs-no-permission"}
2023-09-27T03:37:05Z	INFO	AutoscalingRunnerSet	Skipping cleaning up no permission service account	{"autoscalingrunnerset": "github-runners/arc-runner-set-sample", "reason": "annotation key \"actions.github.com/cleanup-github-secret-name\" not present"}
2023-09-27T03:37:05Z	INFO	AutoscalingRunnerSet	Removing finalizer from manager role binding	{"autoscalingrunnerset": "github-runners/arc-runner-set-sample", "name": "arc-runner-set-sample-gha-rs-manager"}
2023-09-27T03:37:05Z	INFO	AutoscalingRunnerSet	Manager role binding has already been deleted	{"autoscalingrunnerset": "github-runners/arc-runner-set-sample", "name": "arc-runner-set-sample-gha-rs-manager"}
2023-09-27T03:37:05Z	INFO	AutoscalingRunnerSet	Removing finalizer from manager role	{"autoscalingrunnerset": "github-runners/arc-runner-set-sample", "name": "arc-runner-set-sample-gha-rs-manager"}
2023-09-27T03:37:05Z	INFO	AutoscalingRunnerSet	Manager role has already been deleted	{"autoscalingrunnerset": "github-runners/arc-runner-set-sample", "name": "arc-runner-set-sample-gha-rs-manager"}
2023-09-27T03:37:05Z	INFO	AutoscalingRunnerSet	Removing finalizer	{"autoscalingrunnerset": "github-runners/arc-runner-set-sample"}
2023-09-27T03:37:05Z	ERROR	EphemeralRunner	Failed clean up runner from the service	{"ephemeralrunner": "github-runners/arc-runner-set-b2xp4-runner-g447t", "error": "failed to get actions client for runner: failed to get secret: secrets \"arc-runner-set-gha-rs-github-secret\" not found"}
github.com/actions/actions-runner-controller/controllers/actions%2egithub%2ecom.(*EphemeralRunnerReconciler).cleanupRunnerFromService
	github.com/actions/actions-runner-controller/controllers/actions.github.com/ephemeralrunner_controller.go:282
github.com/actions/actions-runner-controller/controllers/actions%2egithub%2ecom.(*EphemeralRunnerReconciler).Reconcile
	github.com/actions/actions-runner-controller/controllers/actions.github.com/ephemeralrunner_controller.go:98
```

**Comment by ChaitanyaAtchuta5** [comment]
Thank you for the reply @Link-
I think these logs are not related to my error. These are logs that got created when I am playing around to create another runner-scale-set "arc-runner-set-sample". Could you please take a look at latest logs(I mean logs at the end) and the logs of the runner pod logs. 
`[WORKER 2023-09-28 03:00:46Z INFO HostContext] Well known config file 'Telemetry': '/home/runner/_diag/.telemetry'
[WORKER 2023-09-28 03:00:46Z INFO JobRunner] Raising job completed event
[WORKER 2023-09-28 03:00:46Z INFO Worker] Job completed.
[RUNNER 2023-09-28 03:00:46Z ERR  GitHubActionsService] GET request to https://pipelinesghubeus4.actions.githubusercontent.com/Ax9WD8GivrPA1LJ2yQVXDqmDwpXfy2igHsA2O9Al6NqzhF0h2P/_apis/distributedtask/pools/1/messages?sessionId=c7a53909-0d63-45d1-8aab-afc1624c58e9&lastMessageId=1&status=Online&runnerVersion=2.309.0 failed. HTTP Status: Forbidden
[RUNNER 2023-09-28 03:00:47Z ERR  MessageListener] Catch exception during get next message.
[RUNNER 2023-09-28 03:00:47Z ERR  MessageListener] GitHub.DistributedTask.WebApi.AccessDeniedException: Access denied. System:ServiceIdentity;DDDDDDDD-DDDD-DDDD-DDDD-DDDDDDDDDDDD needs View permissions to perform the action.
[RUNNER 2023-09-28 03:00:47Z ERR  MessageListener]    at GitHub.Services.WebApi.VssHttpClientBase.HandleResponseAsync(HttpResponseMessage response, CancellationToken cancellationToken)
[RUNNER 2023-09-28 03:00:47Z ERR  MessageListener]    at GitHub.Services.WebApi.VssHttpClientBase.SendAsync(HttpRequestMessage message, HttpCompletionOption completionOption, Object userState, CancellationToken cancellationToken)
[RUNNER 2023-09-28 03:00:47Z ERR  MessageListener]    at GitHub.Services.WebApi.VssHttpClientBase.SendAsync[T](HttpRequestMessage message, Object userState, CancellationToken cancellationToken)
[RUNNER 2023-09-28 03:00:47Z ERR  MessageListener]    at GitHub.Services.WebApi.VssHttpClientBase.SendAsync[T](HttpMethod method, IEnumerable`1 additionalHeaders, Guid locationId, Object routeValues, ApiResourceVersion version, HttpContent content, IEnumerable`1 queryParameters, Object userState, CancellationToken cancellationToken)
[RUNNER 2023-09-28 03:00:47Z ERR  MessageListener]    at GitHub.Runner.Listener.MessageListener.GetNextMessageAsync(CancellationToken token)
[RUNNER 2023-09-28 03:00:47Z INFO MessageListener] Non-retriable exception: Access denied. System:ServiceIdentity;DDDDDDDD-DDDD-DDDD-DDDD-DDDDDDDDDDDD needs View permissions to perform the action.
[RUNNER 2023-09-28 03:00:47Z INFO JobDispatcher] Shutting down JobDispatcher. Make sure all WorkerDispatcher has finished.
[RUNNER 2023-09-28 03:00:47Z INFO JobDispatcher] Ensure WorkerDispather for job 1e5ddf1f-6cd7-5f53-f1c1-e97fb9d5fbae run to finish, cancel any running job.
[RUNNER 2023-09-28 03:00:47Z INFO JobDispatcher] Send job cancellation message to worker for job 1e5ddf1f-6cd7-5f53-f1c1-e97fb9d5fbae.
[RUNNER 2023-09-28 03:00:47Z INFO ProcessChannel] Sending message of length 0, with hash 'e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855'
[RUNNER 2023-09-28 03:00:47Z INFO JobNotification] Entering JobCompleted Notification
[RUNNER 2023-09-28 03:00:47Z INFO JobNotification] Entering EndMonitor`

This the specific error that I am getting from runner pod.

**Comment by ChaitanyaAtchuta5** [comment]
@Link- Here are the short controller logs for the same issue
**Controller log**
`2023-09-29T01:45:25Z	INFO	EphemeralRunnerSet	Scaling comparison	{"ephemeralrunnerset": "github-runners/arc-runner-set-ws992", "current": 1, "desired": 1}
2023-09-29T01:45:25Z	INFO	AutoscalingRunnerSet	Find existing ephemeral runner set	{"autoscalingrunnerset": "github-runners/arc-runner-set", "name": "arc-runner-set-ws992", "specHash": "6b77f84ddc"}
2023-09-29T01:45:33Z	INFO	EphemeralRunner	Ephemeral runner container is still running	{"ephemeralrunner": "github-runners/arc-runner-set-ws992-runner-gx56b"}
2023-09-29T01:45:55Z	INFO	EphemeralRunnerSet	Ephemeral runner counts	{"ephemeralrunnerset": "github-runners/arc-runner-set-ws992", "pending": 0, "running": 1, "finished": 0, "failed": 0, "deleting": 0}
2023-09-29T01:45:55Z	INFO	EphemeralRunnerSet	Scaling comparison	{"ephemeralrunnerset": "github-runners/arc-runner-set-ws992", "current": 1, "desired": 1}
2023-09-29T01:45:55Z	INFO	EphemeralRunner	Ephemeral runner container is still running	{"ephemeralrunner": "github-runners/arc-runner-set-ws992-runner-gx56b"}
2023-09-29T01:46:52Z	INFO	EphemeralRunnerSet	Ephemeral runner counts	{"ephemeralrunnerset": "github-runners/arc-runner-set-ws992", "pending": 0, "running": 1, "finished": 0, "failed": 0, "deleting": 0}
2023-09-29T01:46:52Z	INFO	EphemeralRunnerSet	Scaling comparison	{"ephemeralrunnerset": "github-runners/arc-runner-set-ws992", "current": 1, "desired": 0}
2023-09-29T01:46:52Z	INFO	EphemeralRunnerSet	Deleting ephemeral runners (scale down)	{"ephemeralrunnerset": "github-runners/arc-runner-set-ws992", "count": 1}
2023-09-29T01:46:52Z	INFO	actions-clients	retrieve actions client	{"githubConfigURL": "https://github.com/ORGANIZATION/sample-github-runner", "namespace": "github-runners"}
2023-09-29T01:46:52Z	INFO	AutoscalingRunnerSet	Find existing ephemeral runner set	{"autoscalingrunnerset": "github-runners/arc-runner-set", "name": "arc-runner-set-ws992", "specHash": "6b77f84ddc"}
2023-09-29T01:46:52Z	INFO	actions-clients	using cache client	{"githubConfigURL": "https://github.com/ORGANIZATION/sample-github-runner", "namespace": "github-runners"}
2023-09-29T01:46:52Z	INFO	EphemeralRunnerSet	Skipping ephemeral runner since it is running a job	{"ephemeralrunnerset": "github-runners/arc-runner-set-ws992", "name": "arc-runner-set-ws992-runner-gx56b", "jobRequestId": 312}`

The controller is mentioning that it is skipping the ephemeral runner since it is running a job, but actually there is no job running from my github repository it has completed long back. 

Here is the screenshot of job completion status:

![image](https://github.com/actions/actions-runner-controller/assets/68017180/894dc4f6-a935-4560-81c7-a3b7fdbe130d)

<img width="1670" alt="image" src="https://github.com/actions/actions-runner-controller/assets/68017180/b212d300-acfa-4d21-8743-078c7c72c6fe">


Screenshot of runner pod keeps running even though the job got completed

<img width="1109" alt="image" src="https://github.com/actions/actions-runner-controller/assets/68017180/3d7ac78c-b4c0-48fd-9424-a1a82b3ca3ae">

This runner pod "arc-runner-set-ws992-runner-gx56b" has been there for 8hrs and its still not terminated.

**Comment by Link-** [comment]
@ChaitanyaAtchuta5 - share the controller and listener logs in full (from setup to hitting the issue) please. Share them as a gist not here. Is the problem intermittent or reproducible? If reproducible, what are the exact steps?

**Comment by ChaitanyaAtchuta5** [comment]
@Link- The problem is intermittent but happening almost 80% of the time. 
As mentioned before, the issue is runner pod not getting terminated when the job gets completed. I'll do one thing, I'll remove both the helm deployments(gha-runner-scale-set-controller & gha-runner-scale-set). Install them freshly, reproduce the error and capture the logs.

**Comment by ChaitanyaAtchuta5** [comment]
@Link- I deleted the helm deployments of gha-runner-scale-set-controller & gha-runner-scale-set and installed them again freshly.
**Note:**
1. I am not using any containerMode, just a normal mode with this runner image "ghcr.io/actions/actions-runner:latest"
2. I haven't set any min or max runners
3. Using pat token at repository level
4. AKS cluster with isto-proxy setup

After they got installed without any errors and made sure both the controller and listener pods were up and running. I have triggered a workflow. The new runner pod came up and completed the job. Then he runner pod went to terminating stage and got terminated. Suddenly a new runner pod got into init stage at this point of time there are no workflows running in my repo. After few seconds that runner pod also got terminated. Everything looks okay for now.
Later, I went a head and re-ran the same workflow, the runner pod came up and completed the job as usual. But this time the runner pod was not terminated even after the job was completed, which is not an expected behavior.

**Workflow status screenshot**
<img width="1674" alt="image" src="https://github.com/actions/actions-runner-controller/assets/68017180/c25ae92d-f24f-4b63-8b88-bec6182b9dc9">

**Runner-set status from github**
<img width="342" alt="image" src="https://github.com/actions/actions-runner-controller/assets/68017180/0103d593-ffd4-493b-80a3-9289bd6aa46a">

**Kubectl output**
<img width="1136" alt="image" src="https://github.com/actions/actions-runner-controller/assets/68017180/dcf91afa-70eb-4f5a-9daf-34dc31c9cfc2">

**Controller Logs (from setup to hitting the issue)**
https://gist.github.com/ChaitanyaAtchuta5/b579e55b710b6e98f0760b70442cbd7a

**Listner logs (from setup to hitting the issue)**
https://gist.github.com/ChaitanyaAtchuta5/488ee8b8c1dbe09d9d40bedea243b859

**Runner pod logs that not got terminated**
https://gist.github.com/ChaitanyaAtchuta5/9bff813e8be3b6a4dab655115c9582dc

**Comment by Link-** [comment]
Thank you @ChaitanyaAtchuta5 this is very helpful, we'll take a look!

**Comment by nikola-jokic** [comment]
Hey @ChaitanyaAtchuta5,

The command field might be the cause of the issue. When you provide a command, it serves as an entrypoint. The signal handling is then left to the bash shell to handle the SIGINT signals. The run.sh should act as an entrypoint, but in this case, it may not be able to react on signals since the PID of the run.sh is not going to be 1. Can you please try specifying the command as follows:
```yaml
command: [ "/bin/bash","-c","sleep 10 && exec /home/runner/run.sh" ]
```

Please let us know if this solves the issue :relaxed:

**Comment by ChaitanyaAtchuta5** [comment]
@nikola-jokic Thank You for your reply. I have tried with the above command and still see the same issue.

**Comment by nikola-jokic** [comment]
Hey @ChaitanyaAtchuta5,

I am failing to reproduce this issue. Can you please check if the issue occurs if you don't provide a proxy? It is possible that proxy blocks some requests and therefore the runner never terminates.

**Comment by ChaitanyaAtchuta5** [comment]
@nikola-jokic - You mean without istio-proxy?

**Comment by nikola-jokic** [comment]
Yes, it may be possible that proxy blocks certain API requests that are influencing the behavior

**Comment by ChaitanyaAtchuta5** [comment]
@nikola-jokic I was able to bypass the injection of istio-proxy sidecar container on controller and runner pod by passing an annotation. But I was not able to pass that annotation to bypass it on listener pod. Can you please let me know if there is any way to pass an annotation to listener pod. 
I have tried passing it manually to "AutoscalingListener/arc-runner-set-xxxxxx-listener" but it was not working. 
Thank You

**Comment by nikola-jokic** [comment]
You can modify annotations in `0.6.1` version by using [listenerTemplate](https://github.com/actions/actions-runner-controller/blob/ce80adb9ab06fe990855f59ab47bc2465b86bb3d/charts/gha-runner-scale-set/values.yaml#L90)

**Comment by ChaitanyaAtchuta5** [comment]
@nikola-jokic Thank you, that was helpful. I was able to bypass the istio-proxy on the listener pod as well. But I still see the same issue. I think it has to do something on with the runner. Every time I see this error on runner logs when ever the pod is not terminated as expected. Can you please look into this error.
[https://gist.github.com/ChaitanyaAtchuta5/a90d5dd8798338fabb92f6493b51968c](https://gist.github.com/ChaitanyaAtchuta5/a90d5dd8798338fabb92f6493b51968c)

**Comment by ChaitanyaAtchuta5** [comment]
Finally - after debugging and trying out multiple options. I was able to find that dynatrace one-agent (used for monitoring/ collecting metrics) init container injection was the cause. When I added an annotation to disable that injection it the runner pods are getting terminated as expected. I have also done PID analysis with and without one-agent injection but I couldn't find the root cause. I would like to know is there anyway that I could still use the one-agent but without runner pod termination problem.
@Link- @nikola-jokic Thank you for your help
